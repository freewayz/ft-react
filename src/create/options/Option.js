import React from 'react'
import Answer from './Answer'
import FeedBack from './FeedBack'

class Option extends React.Component {

	state = {
		is_correct: this.props.option.is_correct
	}

	handleIsCorrect = () => {
		const checked = !this.state.is_correct
		this.setState(({is_correct}) => ({is_correct: checked}))
		this.optionChange({is_correct: checked})

	}

	optionChange = (value) => {
		// update redux store with any of children 
		// component value change (Answer,FeedBack and Checkbox)
		const {id,onUpdate} = this.props
		onUpdate({id, ...value})
	}
	render() {
		const {option} = this.props
		const {is_correct} = this.state
		const commonProps = {
			handleUpdate: this.optionChange
		}
		const panelStyle = is_correct ? 'is__correct': ''

		return (
			<div className='option__panel'>
				<div className={`option__panel__header ${panelStyle}`}>
					<span className='option__label'> A </span>
					<span> {is_correct ? 'Correct' : 'Mark as correct'}
			<input 
							type='checkbox' 
							onChange={this.handleIsCorrect}
							checked={this.state.is_correct} />
					</span>
				</div>
				<div className='option__panel__content'>
					{/* Handle compoenent differently so we can test and refactor with ease */}
				<Answer info={option.answer} handleUpdate={this.optionChange} />
				<FeedBack info={option.feedback} {...commonProps} />
			</div>
			</div>
		)
	}
}

export default Option
