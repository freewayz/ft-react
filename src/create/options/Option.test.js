/* global expect it describe shallow mount jest */
import React from 'react'
import Option from './Option'

const mockFn = jest.fn()

const optionProps = {
  is_correct: false,
  answer: 'This is not correct',
  feedback: 'We are testing'
}

describe('Option Component', () => {
  it('renders the right snapshot', () => {
    const wrapper = shallow(<Option option={optionProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('renders the right content', () => {
    const wrapper = shallow(<Option option={optionProps} />)
    const instance = wrapper.instance()
    expect(instance).not.toBe(null)
    expect(wrapper.find('.option__panel').getElement().type).toEqual('div')
  })

  it('change the panel class when setting correct anser', () => {
    const wrapper = shallow(<Option option={optionProps} onUpdate={mockFn} />)
    const instance = wrapper.instance()
    expect(instance).not.toBe(null)
    expect(wrapper.find('.option__panel').getElement().type).toEqual('div')
    const checkBox = wrapper.find('input').getElement()
    // it should be unchecked when is_correct == false
    expect(checkBox.props.checked).toBe(false)
    // simulate clicking
    wrapper.find('input').simulate('change')
    expect(wrapper.state().is_correct).toBe(true)
  })
})
