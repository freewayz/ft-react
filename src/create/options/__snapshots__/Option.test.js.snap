// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option Component renders the right snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Option
    option={
        Object {
            "answer": "This is not correct",
            "feedback": "We are testing",
            "is_correct": false,
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="option__panel__header "
>
          <span
                    className="option__label"
          >
                     A 
          </span>
          <span>
                     
                    Mark as correct
                    <input
                              checked={false}
                              onChange={[Function]}
                              type="checkbox"
                    />
          </span>
</div>,
        <div
          className="option__panel__content"
>
          <Answer
                    handleUpdate={[Function]}
                    info="This is not correct"
          />
          <FeedBack
                    handleUpdate={[Function]}
                    info="We are testing"
          />
</div>,
      ],
      "className": "option__panel",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="option__label"
>
               A 
</span>,
            <span>
               
              Mark as correct
              <input
                            checked={false}
                            onChange={[Function]}
                            type="checkbox"
              />
</span>,
          ],
          "className": "option__panel__header ",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " A ",
              "className": "option__label",
            },
            "ref": null,
            "rendered": " A ",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                " ",
                "Mark as correct",
                <input
                  checked={false}
                  onChange={[Function]}
                  type="checkbox"
/>,
              ],
            },
            "ref": null,
            "rendered": Array [
              " ",
              "Mark as correct",
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "checked": false,
                  "onChange": [Function],
                  "type": "checkbox",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Answer
              handleUpdate={[Function]}
              info="This is not correct"
/>,
            <FeedBack
              handleUpdate={[Function]}
              info="We are testing"
/>,
          ],
          "className": "option__panel__content",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleUpdate": [Function],
              "info": "This is not correct",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "handleUpdate": [Function],
              "info": "We are testing",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="option__panel__header "
>
            <span
                        className="option__label"
            >
                         A 
            </span>
            <span>
                         
                        Mark as correct
                        <input
                                    checked={false}
                                    onChange={[Function]}
                                    type="checkbox"
                        />
            </span>
</div>,
          <div
            className="option__panel__content"
>
            <Answer
                        handleUpdate={[Function]}
                        info="This is not correct"
            />
            <FeedBack
                        handleUpdate={[Function]}
                        info="We are testing"
            />
</div>,
        ],
        "className": "option__panel",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="option__label"
>
                 A 
</span>,
              <span>
                 
                Mark as correct
                <input
                                checked={false}
                                onChange={[Function]}
                                type="checkbox"
                />
</span>,
            ],
            "className": "option__panel__header ",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " A ",
                "className": "option__label",
              },
              "ref": null,
              "rendered": " A ",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  " ",
                  "Mark as correct",
                  <input
                    checked={false}
                    onChange={[Function]}
                    type="checkbox"
/>,
                ],
              },
              "ref": null,
              "rendered": Array [
                " ",
                "Mark as correct",
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "checked": false,
                    "onChange": [Function],
                    "type": "checkbox",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Answer
                handleUpdate={[Function]}
                info="This is not correct"
/>,
              <FeedBack
                handleUpdate={[Function]}
                info="We are testing"
/>,
            ],
            "className": "option__panel__content",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleUpdate": [Function],
                "info": "This is not correct",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "handleUpdate": [Function],
                "info": "We are testing",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
